compiler/rustc_codegen_ssa/src/codegen_attrs.rs

```diff
@@ -621,6 +623,44 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, did: LocalDefId) -> CodegenFnAttrs {
     }
     check_link_name_xor_ordinal(tcx, &codegen_fn_attrs, link_ordinal_span);
 
+    if codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::RUSTC_STD_INTERNAL_SYMBOL)
+        && codegen_fn_attrs.flags.contains(CodegenFnAttrFlags::NO_MANGLE)
+    {
+        let (span, lang_item) = lang_items::extract(attrs)
+            .map_or((None, None), |(name, span)| (Some(span), LangItem::from_name(name)));
+        // This should never use the default as `no_mangle` is present in the attributes list
+        let no_mangle_span = no_mangle_span.unwrap_or_default();
+        let label = if let Some(lang_item) = lang_item {
+            format!("should be the {} item", lang_item.name())
+        } else {
+            format!("should be the internal language item")
+        };
+        let mut err = tcx
+            .dcx()
+            .struct_span_err(
+                no_mangle_span,
+                "`#[no_mangle]` cannot be used on internal language items",
+            )
+            .with_note("The linker requires specific names for internal language items.")
```

Comment:
```suggestion
            .with_note("Rustc requires this item to have a specific mangled name.")
```

not all of them are lang items, you can also use `#[rustc_std_internal_symbol]` explicitly. And it is rustc which has the expectation on the symbol name, not the linker.